gc.c

```diff
@@ -9632,6 +9632,27 @@ garbage_collect_with_gvl(rb_objspace_t *objspace, unsigned int reason)
     }
 }
 
+static int
+gc_promote_object_i(void *vstart, void *vend, size_t stride, void *data)
+{
+    rb_objspace_t *objspace = &rb_objspace;
+    VALUE v = (VALUE)vstart;
+    for (; v != (VALUE)vend; v += stride) {
+        switch (BUILTIN_TYPE(v)) {
+          case T_NONE:
+          case T_ZOMBIE:
+            break;
+          default:
+            if (!RVALUE_OLD_P(v) && !RVALUE_WB_UNPROTECTED(v)) {
+                RVALUE_AGE_SET(v, RVALUE_OLD_AGE);
+                RVALUE_OLD_UNCOLLECTIBLE_SET(objspace, v);
+            }
```

Comment:
I don't think you can do this. If there's a WB protected object that refers to a WB unprotected object then the WB protected object will be set as old while the WB unprotected object will not be promoted.

We should put the parent object in the remember set if it's WB unprotected. Unfortunately, I think we need to use a visitor (like marking) instead of iterating over all the objects.

Reply:
So you made me think that instead we could set them all to `RVALUE_OLD_AGE - 1` (AKA `SET_CANDIDATE`) first, and then run the major GC after.

I pushed a commit that does that.

Reply:
Yeah I think that's a cleaner solution.